Add support for custom PID in history hash

mwarwick opened the pull request:
Custom PID aliases may be provided in the items passed when the gallery is instantiated (similar to the galleryUID option) and will be carried forward into the history hash accordingly.

To facilitate auto-opening a particular item when the page is loaded containing a hash with a custom PID alias, the getItemAt function in items-controller now detects a NaN parameter and searches the items' pid properties for it. This solution only requires slightly modifying the photoswipeParseHash function in your sample implementation script on the Getting Started doc to, e.g.:

if (!isNaN(params.pid)) {
	params.pid = parseInt(params.pid,10)-1;
	if( params.pid < 0 ) {
		params.pid = 0;
	}
}

... so that a hash with, say, "gid=1&pid=photo12345" will cleanly find and display whichever item in the gallery has the {pid: 'photo12345'} property.

dimsemenov commented:
Thank you.

- What if custom PID is a number? Maybe add some flag to options, instead of checking via `isNaN`.
- Not sure if I understand what's this for: `var searchfor = params.pid; var searchfor = params.pid; params.pid = 0;`.
- Please use `===` instead of `==`.

mwarwick commented:
I like the idea about switching to an option flag to invoke the custom PID behavior.  I will modify to your specs and update the pull.

To your question, the purpose of my changes to _parseItemIndexFromURL is to determine the numeric index within items of the image with the custom PID detected in the hash.  We have already parsed the hash into the params object, so I store a copy of the custom PID value in the temporary "searchfor" var and re-initialize params.pid with a zero index before the search.  This will cause the first item in the gallery to pop up by default in case the custom PID is not found in items.  This restores the behavior already in place which sanitizes params.pid.  That said, I notice that I neglected to compare against "searchfor" in the loop, so mea culpa!

mwarwick commented:
Corrective changes are now in place.  Let me know your thoughts...

kyleder commented:
I didn't look at the outstanding pull requests before submitting my own, but I set about solving the same problem in a slightly different way:

https://github.com/dimsemenov/PhotoSwipe/pull/840

dimsemenov commented:
Thanks guys, I'll try to merge this in next few days.

dimsemenov commented:
@mwarwick, please change two more things if you want the pull request to be merged directly: 

1. If option `galleryPIDs` is enabled and developer provided numeric pid's, there might be problems with function `getItemAt` in `items-controller.js`  (as real index might conflict with pid). Please elaborate a bit on why this modification is required. 
2. In `_parseItemIndexFromURL` function  (`history.js`) variable `i` is defined twice, define it only once before the two `for` loops, e.g.:

```js
var i;

for(i = 0; i < ...; i++) {

}

...

for(i = 0; i < ...; i++) {

} 
```

Thanks.

mwarwick commented:
I have added a commit for the variable declaration in `history.js`.

To your question about changes I recommend to `items-controller.js`, the primary motivation is to support your published methodology on the Getting Started doc under "Vanilla JS implementation with IE8 support." There, you advise users to define a `photoswipeParseHash` function whose output is a `params` object with `pid` as it appears in the location hash. If that is a custom PID, then it will be passed to the `openPhotoSwipe` call, which will pass it through as the `index` property of the `options` object to the `new PhotoSwipe` instantiation. As the `index` option, the custom PID is next referenced when `self.currItem = _getItemAt( _currentItemIndex )` is called in `core.js`. Hence, my recommendation to cleanly detect an item being looked up by its custom PID if that is what is known rather than the numeric index.

Rather than extensively overhaul the "vanilla" code to build front-load PID search capabilities into it and guarantee a numerical index, and the secondary motivation of my recommended change to `items-controller.js` being the convenience of passing a custom PID to jump to and make the plugin "do" more with the custom PID extensibility, I modified `getItemAt` to work with a PID just as readily as an item index, since that is (at least for my use case) the preferable and most intuitive way of referring to a particular image. And if a user is not using custom PIDs anyway, then `getItemAt` works just as it does now.

dimsemenov commented:
Thanks for the explanation. But it doesn't solve the conflict between the real index and `pid`. Function `_getItemAt ` is used internally very often.

For example: 
- there is a gallery with 15 images,
- the third image has PID `12` (e.g. because developer decided to use indexes from the database), 
- let's assume that transition starts from a slide with real index 12 to slide 13
- PhotoSwipe internally calls `_getItemAt(12)` to read data about the next slide
- With your current modification, `_getItemAt` will return the third slide object, but it should return slide object 13 (as real index starts from 0).

I propose:
- Remove all changes from `items-controller.js`.
- Modify code in Getting Started of docs, or add some explanation to the description of `galleryPIDs `. Yes, it's an overhaul, but I don't see another solution for now. I'll do this part by myself after your pull request is accepted.

I don't think that it's a good idea to add function like `getItemByPID` to `history.js`, as it'll be used quite rarely. Do you?

Also, please clean up your pull request, so it contains just one or two commits to `history.js`.  You may modify this one, or open new one.

mwarwick commented:
No problem. I'll agree to disagree about the utility and deftness of my solution and whether it impacts internal calls in any meaningful way. To be clear, I do admit the problem if PIDs collide with item indexes, and a utility function for the purpose seems like overkill. I will clean up my pull so you can move forward as desired with commits to `history.js` and will look forward to your handling in Getting Started. Perhaps your vanilla code could make use of `_parseItemIndexFromURL` directly if the method gets exposed rather than writing a custom function to parse the hash?

Also feel free to accept kyleder's fork. I see potential in his approach, though it arguably requires a thorough example. I had to think about it for a while, and it would probably help a lot if his method names were perhaps more intuitive.
